packages/angular/ssr/src/i18n.ts

```diff
@@ -43,3 +43,147 @@ export function getPotentialLocaleIdFromUrl(url: URL, basePath: string): string
+ * @param supportedLocales - An array of supported locales (e.g., `['en-US', 'fr-FR']`),
+ *                           representing the locales available in the application.
+ * @returns The best matching locale from the supported languages, or `undefined` if no match is found.
+ *
+ * @example
+ * ```js
+ * getPreferredLocale('en-US;q=0.8,fr-FR;q=0.9', ['en-US', 'fr-FR', 'de-DE'])
+ * // returns 'fr-FR'
+ *
+ * getPreferredLocale('en;q=0.9,fr-FR;q=0.8', ['en-US', 'fr-FR', 'de-DE'])
+ * // returns 'en-US'
+ *
+ * getPreferredLocale('es-ES;q=0.7', ['en-US', 'fr-FR', 'de-DE'])
+ * // returns undefined
+ * ```
+ */
+export function getPreferredLocale(
+  header: string,
+  supportedLocales: ReadonlyArray<string>,
+): string | undefined {
+  const parsedLocales = parseLanguageHeader(header);
+
+  // Handle edge cases:
+  // - No preferred locales provided.
+  // - Only one supported locale.
+  // - Wildcard preference.
+  if (
+    parsedLocales.size === 0 ||
+    supportedLocales.length < 2 ||
+    (parsedLocales.size === 1 && parsedLocales.has('*'))
+  ) {
+    return supportedLocales[0];
+  }
+
+  // Create a map for case-insensitive lookup of supported locales.
+  // Keys are normalized (lowercase) locale values, values are original casing.
+  const normalizedSupportedLocales = new Map<string, string>();
+  for (const locale of supportedLocales) {
+    normalizedSupportedLocales.set(locale.toLowerCase(), locale);
+  }
+
+  // Iterate through parsed locales in descending order of quality.
+  let bestMatch: string | undefined;
+  const qualityZeroNormalizedLocales = new Set<string>();
+  for (const [locale, quality] of parsedLocales) {
+    const normalizedLocale = locale.toLowerCase();
+    if (quality === 0) {
+      qualityZeroNormalizedLocales.add(normalizedLocale);
+      continue; // Skip locales with quality value of 0.
+    }
+
+    // Exact match found.
+    if (normalizedSupportedLocales.has(normalizedLocale)) {
+      return normalizedSupportedLocales.get(normalizedLocale);
+    }
+
+    // If an exact match is not found, try prefix matching (e.g., "en" matches "en-US").
+    // Store the first prefix match encountered, as it has the highest quality value.
+    if (bestMatch !== undefined) {
+      continue;
+    }
+
+    const [languagePrefix] = normalizedLocale.split('-', 1);
+    for (const supportedLocale of normalizedSupportedLocales.keys()) {
+      if (supportedLocale.startsWith(languagePrefix)) {
+        bestMatch = normalizedSupportedLocales.get(supportedLocale);
+        break; // No need to continue searching for this locale.
+      }
+    }
+  }
+
+  if (bestMatch !== undefined) {
+    return bestMatch;
+  }
+
+  // Return the first locale that is not quality zero.
+  for (const [normalizedLocale, locale] of normalizedSupportedLocales) {
+    if (!qualityZeroNormalizedLocales.has(normalizedLocale)) {
+      return locale;
+    }
+  }
```

Comment:
Question: I've never really understood, what exactly are the semantics of `q=0` anyways? Why would a browser send that? How is it different from just not sending a locale?

Based on this implementation, it seems like we're interpreting that to mean "I do not want this locale" and we therefore pick any other locale the user didn't explicitly reject. Is that the intended usage?

Reply:
Yes, a quality value of 0 means the locale is explicitly not acceptable.